client/grpc_query.go

```diff
@@ -55,56 +56,64 @@ func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, req, reply i
 		return err
 	}
 
-	reqBz, err := ctx.gRPCCodec().Marshal(req)
-	if err != nil {
-		return err
-	}
-
-	// parse height header
-	md, _ := metadata.FromOutgoingContext(grpcCtx)
-	if heights := md.Get(grpctypes.GRPCBlockHeightHeader); len(heights) > 0 {
-		height, err := strconv.ParseInt(heights[0], 10, 64)
+	if ctx.GRPCClient == nil {
+		// Case 2-1. Querying state via abci query.
+		reqBz, err := ctx.gRPCCodec().Marshal(req)
 		if err != nil {
 			return err
 		}
-		if height < 0 {
-			return sdkerrors.Wrapf(
-				sdkerrors.ErrInvalidRequest,
-				"client.Context.Invoke: height (%d) from %q must be >= 0", height, grpctypes.GRPCBlockHeightHeader)
-		}
-
-		ctx = ctx.WithHeight(height)
-	}
 
-	abciReq := abci.RequestQuery{
-		Path:   method,
-		Data:   reqBz,
-		Height: ctx.Height,
-	}
+		// parse height header
+		md, _ := metadata.FromOutgoingContext(grpcCtx)
+		if heights := md.Get(grpctypes.GRPCBlockHeightHeader); len(heights) > 0 {
+			height, err := strconv.ParseInt(heights[0], 10, 64)
+			if err != nil {
+				return err
+			}
+			if height < 0 {
+				return sdkerrors.Wrapf(
+					sdkerrors.ErrInvalidRequest,
+					"client.Context.Invoke: height (%d) from %q must be >= 0", height, grpctypes.GRPCBlockHeightHeader)
+			}
+
+			ctx = ctx.WithHeight(height)
+		}
 
-	res, err := ctx.QueryABCI(abciReq)
-	if err != nil {
-		return err
-	}
+		abciReq := abci.RequestQuery{
+			Path:   method,
+			Data:   reqBz,
+			Height: ctx.Height,
+		}
 
-	err = ctx.gRPCCodec().Unmarshal(res.Value, reply)
-	if err != nil {
-		return err
-	}
+		res, err := ctx.QueryABCI(abciReq)
+		if err != nil {
+			return err
+		}
 
-	// Create header metadata. For now the headers contain:
-	// - block height
-	// We then parse all the call options, if the call option is a
-	// HeaderCallOption, then we manually set the value of that header to the
-	// metadata.
-	md = metadata.Pairs(grpctypes.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10))
-	for _, callOpt := range opts {
-		header, ok := callOpt.(grpc.HeaderCallOption)
-		if !ok {
-			continue
+		err = ctx.gRPCCodec().Unmarshal(res.Value, reply)
+		if err != nil {
+			return err
 		}
 
-		*header.HeaderAddr = md
+		// Create header metadata. For now the headers contain:
+		// - block height
+		// We then parse all the call options, if the call option is a
+		// HeaderCallOption, then we manually set the value of that header to the
+		// metadata.
+		md = metadata.Pairs(grpctypes.GRPCBlockHeightHeader, strconv.FormatInt(res.Height, 10))
+		for _, callOpt := range opts {
+			header, ok := callOpt.(grpc.HeaderCallOption)
+			if !ok {
+				continue
+			}
+
+			*header.HeaderAddr = md
+		}
+	} else {
+		// Case 2-2. Invoke grpc.
+		if err = ctx.GRPCClient.Invoke(grpcCtx, method, req, reply, opts...); err != nil {
```

Comment:
Do these requests actually go through ABCI?

Reply:
they should not